pallets/capacity/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+	}
+
+	/// Sets targets Capacity.
+	pub fn set_capacity_for(
+		target: MessageSourceId,
+		capacity_details: CapacityDetails<BalanceOf<T>, T::EpochNumber>,
+	) {
+		CapacityLedger::<T>::insert(target, capacity_details);
+	}
+
+	/// Calculates Capacity.
+	fn calculate_capacity(amount: BalanceOf<T>) -> BalanceOf<T> {
+		amount
+	}
+
+	/// Decrease a staking account's active token and create an unlocking chunk to be thawed at some future block.
+	fn decrease_active_staking_balance(
+		unstaker: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let mut staking_account =
+			Self::get_staking_account_for(unstaker).ok_or(Error::<T>::StakingAccountNotFound)?;
+		ensure!(amount <= staking_account.active, Error::<T>::AmountToUnstakeExceedsAmountStaked);
+
+		let current_epoch: T::EpochNumber = Self::get_current_epoch();
+		let thaw_at =
+			current_epoch.saturating_add(T::EpochNumber::from(T::UnstakingThawPeriod::get()));
+
+		let unstake_result = staking_account.withdraw(amount, thaw_at)?;
+
+		Self::set_staking_account(&unstaker, &staking_account);
+
+		Ok(unstake_result)
+	}
+
+	/// Reduce available capacity of target and return the amount of capacity reduction.
+	fn reduce_capacity(
+		unstaker: &T::AccountId,
+		target: MessageSourceId,
+		amount: BalanceOf<T>,
+	) -> Result<BalanceOf<T>, DispatchError> {
+		let mut staking_target_details = Self::get_target_for(&unstaker, &target)
+			.ok_or(Error::<T>::StakerTargetRelationshipNotFound)?;
+		let mut capacity_details =
+			Self::get_capacity_for(target).ok_or(Error::<T>::TargetCapacityNotFound)?;
+		let capacity_reduction = Self::calculate_capacity_reduction(
+			amount,
+			capacity_details.total_tokens_staked,
+			capacity_details.total_capacity_issued,
+		);
+		staking_target_details.withdraw(amount, capacity_reduction);
+		capacity_details.withdraw(capacity_reduction, amount);
+
+		Self::set_capacity_for(target, capacity_details);
+		Self::set_target_details_for(unstaker, target, staking_target_details);
+
+		Ok(capacity_reduction)
+	}
+
+	/// Determine the capacity reduction when given total_capacity, unstaking_amount, and total_amount_staked.
+	fn calculate_capacity_reduction(
+		unstaking_amount: BalanceOf<T>,
+		total_amount_staked: BalanceOf<T>,
+		total_capacity: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		Perbill::from_rational(unstaking_amount, total_amount_staked).mul_ceil(total_capacity)
+	}
+
+	fn start_new_epoch_if_needed(current_block: T::BlockNumber) -> Weight {
+		if Self::get_current_epoch_info()
+			.epoch_start
+			.saturating_add(Self::get_epoch_length())
+			.eq(&current_block)
+		{
+			let current_epoch = Self::get_current_epoch();
+			CurrentEpoch::<T>::set(current_epoch.saturating_add(1u32.into()));
+			CurrentEpochInfo::<T>::set(EpochInfo { epoch_start: current_block });
+			T::WeightInfo::on_initialize()
+				.saturating_add(RocksDbWeight::get().reads(1))
+				.saturating_add(RocksDbWeight::get().writes(1))
```

Comment:
This still needs to count two writes because we are setting two storage items above, line 578, 579

Reply:
Addressed here.
https://github.com/LibertyDSNP/frequency/pull/1372